scripts/travis/upload-to-registry.sh

```diff
@@ -0,0 +1,112 @@
+#!/bin/bash
+
+set -euxf -o pipefail
+
+DOCKERHUB_USERNAME=${DOCKERHUB_USERNAME:-"jaegertracingbot"}
+DOCKERHUB_TOKEN=${DOCKERHUB_TOKEN:-}
+QUAY_USERNAME=${QUAY_USERNAME:-"jaegertracingbot+github_workflows"}
+QUAY_TOKEN=${QUAY_TOKEN:-}
+
+usage() {
+  echo $"Usage: $0 <image>"
+  exit 1
+}
+
+check_args() {
+  if [ ! $# -eq 1 ]; then
+    echo "ERROR: need exactly one argument"
+    usage
+  fi
+}
+
+compute_image_tag() {
+  local branch=$1
+
+  if [[ "${branch}" == "master" ]]; then
+    local tag="latest"
+    echo ${tag}
+  elif [[ "${branch}" =~ ^v([0-9]+)\.([0-9]+)\.([0-9]+)$ ]]; then
+    local major="${BASH_REMATCH[1]}"
+    local minor="${BASH_REMATCH[2]}"
+    local patch="${BASH_REMATCH[3]}"
+    local tag=${major}.${minor}.${patch}
+    echo "${tag} ${major} ${minor} ${patch}"
+  else
+    local tag="${branch}"
+    echo ${tag}
+  fi
+}
+
+label_release_tag() {
+  local registry=$1
+  local image=$2
+  local major=${3:-}
+  local minor=${4:-}
+  local patch=${5:-}
+
+  if [[ -n ${major} ]]; then
+    docker tag ${image} ${registry}/${image}:${major}
+    if [[ -n ${minor} ]]; then
+      docker tag ${image} ${registry}/${image}:${major}.${minor}
+      if [[ -n ${patch} ]]; then
+        docker tag ${image} ${registry}/${image}:${major}.${minor}.${patch}
+      fi
+    fi
+  fi
+}
+
+try_login() {
+  local registry=$1
+  local user=$2
+  local token=$3
+  local marker=$4
+
+  if [ ! -f ${marker} ] && [ -v ${token} ]; then
+    printenv ${token} | docker login ${registry} --username ${user} --password-stdin
+    touch ${marker}
+  fi
+}
+
+upload_images() {
+  local registry=$1
+  local image=$2
+  local user=$3
+  local token=$4
+  local marker=.${registry}.login
+
+  try_login ${registry} ${user} ${token} ${marker}
+
+  if [ ! -f ${marker} ] && [ -z ${token} ]; then
```

Comment:
another way to refactor this is to do something like this:
```
if [ try_login ${registry} ${user} ${token} ] ; then
    # do upload
else
    # echo no upload
fi
```

Reply:
This way the marker can be hidden inside try_login

Reply:
Trying it.

Reply:
I am little confused about this change. If we use this way, let's say we don't specify token for one of the registries, won't it will still try to upload to that registry since we no longer checking for marker file and deciding based on the return status of try_login status.

Reply:
In this specific location, if the marker file does not exist, it means we did not log in, so no point in trying to upload. That's why the extra `&& -v $token` condition was wrong.

Otherwise, if you're asking about the `if [ try_login ]` version, then the function should return true only if it successfully logged in, i.e. it needs to be modified slightly to return false when token is not defined.

I wouldn't worry about it, lets leave it as it is now.